batchflow/models/torch/callbacks/base.py

```diff
@@ -0,0 +1,135 @@
+""" Base class for model callbacks: describes API. """
+from abc import ABC, abstractmethod
+from time import gmtime, strftime
+
+from ....monitor import USSMonitor, GPUMonitor, GPUMemoryMonitor
+
+
+def file_print(path, msg):
+    """ Print to a file. """
+    with open(path, 'a+') as f:
+        print(msg, file=f)
+
+
+
+class BaseCallback(ABC):
+    """ Base class for callbacks.
+
+    Parameters
+    ----------
+    stream : None, callable or str
+        If None, then no logging is performed.
+        If callable, then used to display message, for example, `print`.
+        If str, then must be path to file to write log to.
+    """
+    def __init__(self, stream=None):
+        self.model = None
+        self.stream = self.make_stream(stream)
+
+    def make_stream(self, stream):
+        """ Parse the `stream` argument into callable. """
+        if stream is None:
+            return lambda *args: args
+
+        if callable(stream):
+            return stream
+
+        if isinstance(stream, str):
+            with open(stream, 'w') as _:
+                pass
+            return lambda msg: file_print(stream, msg)
+        raise TypeError('`Stream` argument must be either None, callable or string.')
+
+    def set_model(self, model):
+        """ Save reference to the model.
+        Depending on the callback, following attribytes may be accessed:
+            - `iteration`
+            - `loss_list`
+            - `microbatch`
+            - `iter_info`
+        """
+        self.model = model
+        return self
+
+    @abstractmethod
+    def on_iter_end(self, **kwargs):
+        """ Action to be performed on the end of iteration. Called at the end of :meth:`TorchModel.train`. """
+
+
+class LogCallback(BaseCallback):
+    """ Log loss and various model information at the desired stream.
+
+    Parameters
+    ----------
+    stream : None, callable or str
+        If None, then no logging is performed.
+        If callable, then used to display message, for example, `print`.
+        If str, then must be path to file to write log to.
+    frequency : int
+        Frequency of log messages.
+    resources : bool
+        Whether to log memory, GPU memory and GPU utilization.
+    shapes : bool
+        Whether to log model input and output shapes.
+    microbatch : bool
+        Whether to log size and number of microbatches.
+    """
+    def __init__(self, stream=None, frequency=1, resources=True, shapes=False, microbatch=False):
+        self.frequency = frequency
+        self.resources = resources
+        self.shapes = shapes
+        self.microbatch = microbatch
+
+        self.header = self.make_header()
```

Comment:
do you really need `header` to be an attribute? will you need it after writing it to the log?